src/edge_index.cpp

```diff
@@ -106,6 +111,127 @@ Selection resolve(const HighFive::Group& indexGroup, const std::vector<NodeID>&
     return Selection::fromValues(result);
 }
 
+
+namespace {
+
+std::unordered_map<NodeID, RawIndex> _groupNodeRanges(const std::vector<NodeID>& nodeIDs)
+{
+    std::unordered_map<NodeID, RawIndex> result;
+
+    if (nodeIDs.empty()) {
+        return result;
+    }
+
+    uint64_t i0 = 0;
+    NodeID lastNodeID = nodeIDs[0];
+    for (uint64_t i = 1; i < nodeIDs.size(); ++i) {
+        if (nodeIDs[i] != lastNodeID) {
+            result[lastNodeID].push_back({i0, i});
+            i0 = i;
+            lastNodeID = nodeIDs[i0];
+        }
+    }
+
+    result[lastNodeID].push_back({i0, nodeIDs.size()});
+
+    return result;
+}
+
+
+std::vector<NodeID> _readNodeIDs(const HighFive::Group& h5Root, const std::string& name)
+{
+    std::vector<NodeID> result;
+    h5Root.getDataSet(name).read(result);
+    return result;
+}
+
+
+void _writeIndexDataset(const RawIndex& data, const std::string& name, HighFive::Group& h5Group)
+{
+    auto dset = h5Group.createDataSet<uint64_t>(name, HighFive::DataSpace::From(data));
+    dset.write(data);
+}
+
+
+void _writeIndexGroup(const std::vector<NodeID>& nodeIDs, NodeID maxNodeID, HighFive::Group& h5Root, const std::string& name)
+{
+    auto indexGroup = h5Root.createGroup(name);
+
+    auto nodeToRanges = _groupNodeRanges(nodeIDs);
+    const auto rangeCount = std::accumulate(
+        nodeToRanges.begin(), nodeToRanges.end(), uint64_t(0),
+        [](uint64_t total, const std::pair<NodeID, RawIndex>& item) {
```

Comment:
`const std::pair<NodeID, RawIndex>&` -> `decltype(nodeToRanges)::const_reference item` ?

Reply:
Thanks for suggesting a C++11 alternative.
I guess it would appeal more to people following modern C++; but might confuse those less savvy (I would apply it nevertheless). ^^